Feature match workflow

opadron opened the pull request:
Implements the feature matching workflow.

Also includes a suite of miscellaneous fixes and new features.  See commit notes for more details.

Commit:
add feature match workflow

Commit:
fix job polling issue

Fixes an issue with job polling where the polling
process would fail to time out, even after having
surpassed the maximum number of tries.

Commit:
add metadata upload feature

Extends the uploadFile() global action creator to
accept additional metadata to apply to the newly
created file after upload.

Commit:
fix addAnalysisElement

Fixes an issue with the addAnalysisElement()
global action creator that caused the wrong
element document to be returned in the case where
the element to be added has its own child
elements.

Commit:
rework triggerAnalysisAction

Reworks the triggerAnalysisAction() global action
creator so that it accepts a single object of
named arguments instead of several positional
arguments.

Also fixes a minor issue in the onAction()
callback passed down to analysis element
components where additional arguments beyond the
action string where being silently ignored.

Commit:
add ensureDirectory action creators

Adds several new global action creators:

 - ensureUserDirectory()

   - checks for a given directory in a user's home
     space and creates it if not found

 - ensureScratchDirectory()

   - works like ensureUserDirectory(), but
     specifically for a particular directory that
     is meant to be used as a temporary staging
     area for intermediate computational results.

 - ensurePrivateDirectory()

   - works like ensureUserDirectory(), but
     specifically for the user's "Private"
     directory.  This directory serves as the
     designated landing and storage area for
     user-uploaded data.

The specific ensureScratchDirectory() and
ensurePrivateDirectory() creators maintain their
own singletons to ensure that concurrent
asyncronous calls do not cause multiple
directories to be created.  The more general
ensureUserDirectory() creator does not and is
therefore not async-safe (or whatever the
javascript equivalent to thread-safe is).  Call
sites must take their own measures to ensure that
multiple concurrent calls do not cause multiple
directories to be created.

Commit:
add populateFileSelectionElement action creator

Adds the convenience global action creator,
populateFileSelectionElement(), that can be used
to programmatically set the state of a
file-selection or folder-selection analysis
element given a girder item or folder,
respectively.

Commit:
fix tabGroupElement

Fixes an issue with the tabGroup analysis element
type that caused rendering to fail when given no
child elements.

Commit:
switch from a dash to an X in the upload UI

Switches to using an X within the "remove" buttons
for each added upload entry instead of a dash in
an effort to better communicate the intent of the
button.

Commit:
fix reloading issue in upload UI

Fixes an issue in the upload UI where clicking on
the "remove" button for a previously-added entry
would sometimes cause the page to reload.

Commit:
fix feature match sniffer code

Fixes a minor issue where the SNIFFER_DELIMITERS
constant was not being used.

Commit:
fix registerAnalysisAction issue

Fixes an issue with the registerAnalysisAction()
global action creator that prevented it from
recognizing when a page was passed directly
instead of its key.

Commit:
add feature extract script

Commit:
simplify feature match workflow

Changes the feature match workflow to accept a
single match specification string instead of a
comma-separated sequence of such strings.

Also removes some old debugging code.

Commit:
add loadModel utility

Commit:
add apply section of the feature match workflow

Commit:
add feature apply script

Commit:
add onlyFileTypes option to file browsing elements

Adds the "onlyFileTypes" option to fileSelection
analysis elements.  This option is checked against
the "sumoFileType" metadata value.

Also renames what was the "onlyTypes" option to
"onlyDataTypes".  This option is (still) checked
against the "sumoDataType" metadata value.

Commit:
add content upload feature

Extends the uploadFile() global action creator to
allow passing the contents of the new file to be
created directly.

Commit:
add groupElement

Adds a new analysis element type, "group", which
is simply an invisible grouping for subelements.
A "group" analysis element is closely analogous to
a <div> DOM element.

Commit:
add girderItemElement

Adds a new analysis element type, "girderItem",
that allows developers to insert stylized girder
item entries with download and viewing links.

Commit:
add utility module: utils/scatter.jsx

Commit:
add utility promise functions

Adds SCHEDULES, a mapping of scheduling constants.

Adds map(), a special version of the mapping
pattern that allows one to control both
concurrency factor and scheduling policy.

Adds semaphore(), which returns a Promise built on
a callback taking two other callbacks, similar to
new Promise((resolve, reject) => (...)), except
the callback takes "increment" and "decrement"
callbacks.  The promise manages an internal
counter that starts at 0.

  increment(v) updates the current resolution
  value to v and increases the internal counter by
  1.

  decrement() reduces the internal counter by 1.
  If the internal counter reaches 0 as a result of
  calling this function, the returned Promise
  resolves with the most recently set resolution
  value.

Adds delayedSemaphore(), which works similarly to
semaphore(), except its Promise is built on a
configurable delay and a callback that only takes
a single "increment" callback.  The Promise
automatically matches every "increment" call with
an internal "decrement" that triggers after the
configured delay time has elapsed.

Commit:
switch from noStrip parameter to helper function

Further improves the addAnalysisElement fix so
that it does not require a secret third argument.

waxlamp commented on web_client/actions.jsx:
```diff
@@ -72,18 +72,43 @@ const promiseAction = (callback) => (dispatch, getState) => new Promise(
+      return (
+        _noStrip
+
+          ? { ...element }
```
Let's avoid the `?:` operator if we can - it seriously impedes readability.

waxlamp commented on web_client/actions.jsx:
```diff
@@ -857,6 +844,30 @@ export const uploadFile = (file, params=null, parent=null) => promiseAction(
+        req.done(() => { resolve(itemModel); });
+        req.error(() => { reject(new Error()); });
+      })
+
```
Remove blank line

waxlamp commented on web_client/actions.jsx:
```diff
@@ -857,6 +844,30 @@ export const uploadFile = (file, params=null, parent=null) => promiseAction(
+          })
+        )
+      )))
+
```
Remove blank line

waxlamp commented on web_client/components/containers/upload.jsx:
```diff
@@ -104,6 +92,7 @@ const UploadContainer = connect(
 
       (
         dispatch(actions.setUploadModeToUploading())
+
```
Remove blank line

waxlamp commented on web_client/actions.jsx:
```diff
@@ -671,14 +674,51 @@ export const toggleHeaderDropdown = () => promiseAction(
   return promiseAction(
     (dispatch, getState) => {
+      // (data, page=null, action=null, options={}) => {
+      let {
```
Why not destructure the argument directly in the argument list in line 677?

waxlamp commented on web_client/actions.jsx:
```diff
@@ -671,14 +674,51 @@ export const toggleHeaderDropdown = () => promiseAction(
+        let { analysis } = getState();
+
+        if (isNil(objects)) {
+          ({ objects } = analysis);
```
Remove parentheses

opadron replied:
See above.

waxlamp commented on web_client/actions.jsx:
```diff
@@ -671,14 +674,51 @@ export const toggleHeaderDropdown = () => promiseAction(
+        }
+
+        if (isNil(states)) {
+          ({ states } = analysis);
```
Remove parentheses

opadron replied:
See above.

waxlamp commented on web_client/actions.jsx:
```diff
@@ -362,6 +362,44 @@ export const openRegisterDialog = (byRouter = false) => promiseAction(
+  (dispatch, getState) => {
+
+    if (isString(item)) {
+      return (
```
Remove parentheses

waxlamp commented on web_client/actions.jsx:
```diff
@@ -362,6 +362,44 @@ export const openRegisterDialog = (byRouter = false) => promiseAction(
+    if (isString(item)) {
+      return (
+        getResourceFromId(item, 'item')
+        .then((item) => dispatch(populateFileSelectionElement(element, item)))
```
Indentation

waxlamp commented on web_client/actions.jsx:
```diff
@@ -223,6 +223,101 @@ export const enableAnalysisPage = (page) => promiseAction(
+      }).then(({ response: folderList }) => (
+        folderList.length === 0
+
+          ? rest({
```
Don't use `?:`

waxlamp commented on web_client/actions.jsx:
```diff
@@ -223,6 +223,101 @@ export const enableAnalysisPage = (page) => promiseAction(
+      }
+
+      if (isNil(user)) {
+        ({ user } = getState().loginInfo);
```
Remove parentheses

opadron replied:
Needed in this case to disambiguate syntax from that of a new code block.

i.e.:

    { user } = ...;

 fails to parse, as it looks like a code block followed by an `=` operator.

waxlamp replied:
In that case, I'd suggest something more like:

```javascript
user = getState().loginInfo.user;
```

waxlamp commented on web_client/actions.jsx:
```diff
@@ -223,6 +223,101 @@ export const enableAnalysisPage = (page) => promiseAction(
+  } = args;
+
+  return promiseAction(
+    (dispatch, getState) => {
```
There may be a simple way to help debug situations where multiple calls are in flight. We simply lift the declaration of `result` in line 280 below to outside of the `return promiseAction(...)` statement in line 266. Before firing the ajax call in line 280, check to see if `result` is not null; if not, then throw an error or something similar. Otherwise, continue as normal, and append a callback that resets `result` to null.

opadron replied:
tl; dr Pretty soon, none of what I wrote below is going to matter.

So, consider these two points about an "idealized" implementation of this function:

1 - Multiple concurrent calls for the same directory are not erroneous; the point is to ensure that the *same* promise chain is returned to each caller in such a case.  This approach avoids race conditions in a way that doesn't require that every call site be aware of each other.  In fact, the same promise chain would be ideally returned even long after having been resolved (effectively making subsequent calls a noop that immediately returns a promise that resolves immediately).  It would be best if you didn't have to go over the wire to ensure the presence of a directory that you've already checked for in a previous call.

2 - A new promise chain is created for every distinct directory queried -- on the first such call for that directory.  This idealized implementation is basically a memoized function returning a promise chain that is unique to the input directory.

The code being discussed differs from the ideal in that I specifically chose not to memoize it, because doing so would require managing a table of promise chains that would grow, possibly without bound, and in a way very unlikely to be obvious to other developers.  I made exceptions for the user's Private folder and the scratch space folder, since OSUMO specifically needs those two folders, and figured that leaving other call sites for other directories to coordinate among themselves was a pragmatic alternative to caching or full memoization.

In any case, @chaitanya2334 and I actually figured out what was the issue with the R scripts, today, and my plan for addressing that issue will mostly supplant this part of the code.

waxlamp replied:
Gotcha, thanks for the explanation. Original comment withdrawn :)

waxlamp commented on web_client/components/body/analysis/element/group.jsx:
```diff
@@ -0,0 +1,10 @@
+class Group extends React.Component {
+  render () {
+    let { children } = this.props;
+    return (<div>{children}</div>);
```
Remove parentheses

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+
```
Remove blank line

opadron replied:
ok with docstring?

waxlamp replied:
:+1:

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+        delimiters=SNIFFER_DELIMITERS
+    )
+
+
```
Remove blank line

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+with open(input_path_2, 'rU') as input2:
+    list2 = [x for x in ColumnExtractor(input2, dialect2)]
+
+extract_result = (
```
Remove parentheses

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+extract_result = (
+    list(
+    sorted(
+        (
```
Remove parentheses

opadron replied:
SyntaxError: Generator expression must be parenthesized if not sole argument

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+    list2 = [x for x in ColumnExtractor(input2, dialect2)]
+
+extract_result = (
+    list(
```
The formatting for this expression is fairly wonky - maybe use a more standard Python style here.

opadron replied:
I'm going to assume by "standard Python style" you mostly mean PEP8, which I'm going to take this chance to cherry pick [quotes](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds) from:

> However, know when to be inconsistent -- sometimes style guide recommendations just aren't applicable. When in doubt, use your best judgment. Look at other examples and decide what looks best. And don't hesitate to ask!
> ...
> Some other good reasons to ignore a particular guideline:
> 1 - When applying the guideline would make the code less readable, even for someone who is used to reading code that follows this PEP.
> ...

I think lengthy expressions like these are good examples of where readability gains from playing things more by ear style-wise could be worth the inconsistency.

Now, the quote *does* call for discussion in cases where it's not obvious, so I'm going to go ahead and take my best shot at PEP8-ing the expression.  If you can honestly say that the result (or another variation you could recommend) is more readable, then we'll go with that.

waxlamp replied:
I of course agree that readability is the most important reason to even have style rules, and as a consequence, I do not generally mean "PEP8" when I say "standard style" (because there are things in PEP8 that I generally do not agree increase readability, and I also believe slavish devotion to a style standard generally causes problems).

In this case, I would probably format the code this way:

```python
values = ({'id': a if len(a) < len(b) else b,
           'description': '{}/{}'.format(a, b)} for a, b in zip(list1, list2))
extract_result = list(sorted(values, key=lambda x: x['id']))
```

Factoring out the generator expression (which is the most complex part of this stretch of code) greatly improves the readability of that second "composition" statement, without having to use non-standard spacing/linebreaks (which was, now that I think about it, my original complaint, rather than non-PEP8ness in general).

The code as you have it is not that bad, but the way you have nested functions indented to the same level, extra parentheses where they're not needed, and a specific linebreak-highlighted use of generator expression parentheses are just not seen in "standard" (by which I mean "typical" rather than "PEP8-compliant") Python code, and that in itself can limit readability.

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+            }
+            for (a, b) in zip(list1, list2)
+        ),
+
```
Remove blank line

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+        key=(lambda x: x['id'])
+    ))
+)
+lastId = None
```
Insert blank line above

waxlamp commented on server/task_specs/feature-extract.py:
```diff
@@ -0,0 +1,117 @@
+        lastId = entry['id']
+
+extract_result = json.dumps(extract_result)
+
```
Remove blank line

waxlamp commented on server/task_specs/feature-apply.py:
```diff
@@ -0,0 +1,147 @@
+
```
Remove blank line

opadron replied:
I think this particular change should be skipped in light of the new docstring?

waxlamp replied:
:+1:

waxlamp commented on server/task_specs/feature-apply.py:
```diff
@@ -0,0 +1,147 @@
+import numpy as np
+from six import PY2
+
+SNIFFER_SIZE = 1024
```
Should this and the following two lines be constants read in from a file or something? Another file has these same constant defined, right?

opadron replied:
Yeah, I was thinking about factoring them out into their own file.  I ended up shelving the idea (for now), but very well may come back to this at a later time.

waxlamp replied:
:+1:

I'd urge us to create an issue for this so we don't forget, and I'd say for such a small change with potential to save us future headaches, we should do it sooner rather than later.

waxlamp commented on server/task_specs/feature-apply.py:
```diff
@@ -0,0 +1,147 @@
+            write_csv_helper(
+                    data_block, row_headers, column_headers, f, dialect)
+
+# # uncomment to test in the command line
```
This commented-out testing block should go below the rest of the module, in a block delimited by something like `if __name__ == '__main__':`

opadron replied:
PTAL and see if you approve.

waxlamp commented on server/task_specs/feature-match.py:
```diff
@@ -0,0 +1,321 @@
+
```
Remove blank line

opadron replied:
ok with docstring?

waxlamp replied:
:+1:

waxlamp commented on server/task_specs/feature-match.py:
```diff
@@ -0,0 +1,321 @@
+
+ROW_SCAN_SIZE = 1024
+
+SNIFFER_SIZE = 1024
```
Constants should move to an external file.

opadron replied:
See above.

waxlamp commented on server/task_specs/feature-match.py:
```diff
@@ -0,0 +1,321 @@
+    if current:
+        result.append(''.join(current))
+
+    return [x for x in result if x]
```
I think you can simply say `return filter(None, result)`

opadron replied:
Nice!

waxlamp commented on server/task_specs/feature-match.py:
```diff
@@ -0,0 +1,321 @@
+    return [x for x in result if x]
+
+
+class ColumnExtractor(object):
```
Is this code being repeated? I think it appeared in a different Python script in this PR, right?

opadron replied:
Yup, see note above.

waxlamp commented on web_client/analysis-modules/base-listing.jsx:
```diff
@@ -3,13 +3,15 @@ import igpse from './igpse';
-  { key: 'kmeans'    , name: 'K Means'        , module: kmeans     },
-  { key: 'silhouette', name: 'Silhouette Plot', module: silhouette },
-  { key: 'surv'      , name: 'Survival Plot'  , module: surv       }
+  { key: 'igpse'     , name: 'iGPSe'           , module: igpse           },
```
Prefer standard Python style for specifying dict literals.

opadron replied:
I think there's an argument to be made similar to my note above about PEP8, but in this case, it seems to me that this code can take on many possible formats and still be readable, so I will go ahead and fix this up.  Please let me know what you think of the result.

Edit: Github is getting wonky on me.  Sorry if this ends up double-posting.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+import { formatSize } from 'girder/misc';
+
+const D = store.dispatch.bind(store);
```
- Don't use capitalized variable names for other than class constructors.
- Prefer descriptive variable names in general.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+    (
+      computeResultElement
+        ? D(actions.removeAnalysisElement(computeResultElement))
```
Avoid `?:`

opadron replied:
See "push back" note.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+    (
+      applyResultElements
+        ? Promise.all(
```
Avoid `?:`

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+      disable: true,
+      remove: false
+    })),
+
```
Remove blank line

opadron replied:
Holding off on this change until I hear your thoughts about my note on the array literals.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+        ? D(actions.removeAnalysisElement(computeResultElement))
+        : null
+    ),
+
```
Remove blank line

opadron replied:
Waiting to hear back...

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+      remove: false
+    })),
+
+    (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+        : null
+    ),
+
+    (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+let matchChooserElement;
+
+const computeMatch = (data, page) => {
+  const truncatePromise = Promise.all([
```
The "list construction" syntax used here makes the code look unusual. Prefer a more standard way to specify the list argument.

opadron replied:
I really can't say that I follow you, here.  What we're looking at is just an array literal (no "'list construction' syntax"), with three elements, formatted in an attempt to make clear where the expression for each element begins and ends.  Clearly, my formatting attempt wasn't entirely successful, but without actionable feedback, I struggle to identify anything concrete worth changing.  The only thing I can think of might be to add an extra level of indentation just for the square brackets to make them stand out more?  I'm really grasping at straws, here.

I think it's also worth pointing out that for function calls with a single array literal argument, I've usually taken the liberty of keeping the array brackets and parentheses together, as in:

    myFunction([
        my,
        array,
        literal
    ]);

And in particular, this comes up a _lot_ with `Promise.all()`, so consider that any recommendations you make here might be applicable to _many_ parts of the code spread out through the whole code base.

waxlamp replied:
I should have been more clear with my coined phrase "list construction" - I just meant the style you were using to highlight elements of the array you were constructing.

The main way I would improve this is to use a `const` variable for each of the promises, then deliver them all together afterwards with a simpler `Promise.all()` expression:

```javascript
const fooPromise = ...;
const barPromise = ...;
const bazPromise = ...;

const truncatePromise = Promise.all([fooPromise, barPromise, bazPromise]);
```

The drawback is that this declares more variables, but they are declared `const` so the potential damage from doing so is bounded. I think the gain in readability justifies the extra declarations. But again, this is a style thing so rely on your own preferences. As in other discussions, I'll just reiterate that the extra parentheses and linebreaks may generally impede readability because they're not constructions found in typical JS code.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+  const runPromise = (
+    analysisUtils.runTask(task, { inputs, outputs }, { title, maxPolls })
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  const runPromise = (
+    analysisUtils.runTask(task, { inputs, outputs }, { title, maxPolls })
+
+    .then(({ match_result: { data: result } }) => result)
```
Indentation

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+    .then(({ match_result: { data: result } }) => result)
+  );
+
+  return (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  const title = 'feature-match';
+  const maxPolls = 40;
+
+  const runPromise = (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+  return (
+    Promise.all([truncatePromise, runPromise])
+    .then(([, result]) => result)
```
Indentation

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  return (
+    Promise.all([truncatePromise, runPromise])
+    .then(([, result]) => result)
+    .then((result) => {
```
Indentation

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+      });
+      console.log(result);
+
+      return (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+            ]
+          }
+        ))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+            },
+            computeTabElement
+          )),
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+            matchChooserElement, item.attributes
+          ))
+        ]))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+const main = () => (
+  Promise.resolve()
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+const main = () => (
+  Promise.resolve()
+
+  .then(() => D(actions.registerAnalysisAction(
```
Indentation

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  .then(() => D(actions.registerAnalysisAction(
+    'feature-match', 'computeMatch', computeMatch))
+  )
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+    'feature-match', 'computeMatch', computeMatch))
+  )
+
+  .then(() => D(actions.registerAnalysisAction(
```
Indentation

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  .then(() => D(actions.registerAnalysisAction(
+    'feature-match', 'applyMatch', applyMatch))
+  )
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+  )
+
+  .then(() => analysisUtils.fetchAnalysisPage('feature-match'))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+    let applyTabElements = applyTab.elements;
+    delete applyTab.elements;
+
+    return (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+    return (
+      D(actions.addAnalysisPage({ ...page, enabled: false }))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+      D(actions.addAnalysisPage({ ...page, enabled: false }))
+
+      .then((page) => (page1 = page))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+      .then(() => D(actions.addAnalysisElement(tabGroup, page1)))
+      .then((e) => (tabGroupElement = e))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+      .then(() => D(actions.addAnalysisElement(computeTab, tabGroupElement)))
+      .then((e) => (computeTabElement = e))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+
+      .then(() => D(actions.addAnalysisElement(applyTab, tabGroupElement)))
+      .then((e) => (applyTabElement = e))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -0,0 +1,170 @@
+      .then(([e]) => (matchChooserElement = e))
+    );
+  })
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
-const applyMatch = (data, page) => { /* TODO(opadron): */ };
+const applyMatch = (data, page) => {
+  const truncatePromise = Promise.all([
+    D(actions.truncateAnalysisPages(1, {
```
Don't use capitalized, short variable name.

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+
+    (
+      applyResultElements
+        ? D(actions.removeAnalysisElement(applyResultElements))
```
Avoid `?:`

opadron replied:
I'm gonna push back in this specific case.  Dropping the tertiary operators would require initializing (and naming) two new variables and moving their corresponding sub-expressions a bit from where they are actually used -- for what seems to be no purpose other than cutting down on colons and question marks.  Also considering how simple these particular cases are, I'm far from convinced that the suggested change would actually be an improvement.

waxlamp replied:
This is another case of "non-standard" syntax making readability more difficult. In this case, you have a parenthesized expression set off from its immediate context by a blank line. It therefore takes a mental "zoom-out" to realize that it's in fact the second element in a list of promises, and a unique one at that: sometimes there is no promise to execute, which is represented by a null value that passes through to the eventual callback. I think the idea would be better expressed by something like this:

```javascript
  let promises = [
    D(actions.truncateAnalysisPages(1, {
      clear: true,
      disable: true,
      remove: false
    })),
    null
  ];

  if (applyResultElements) {
    promises[1] = D(actions.removeAnalysisElement(applyResultElements);
  } 

  const truncatePromise = Promise.all(promises);
```

This is not the most elegant JavaScript I've ever written, and it's definitely not wrong in an objective sense to use `?:` here, but I still think my version is more readable. It is ultimately a matter of style, so go with what makes the most sense to you.

The only part of my critique I want to absolutely leave in place is that non-standard spacing/linebreaks/syntax will definitely slow down future readers of this code, and should be avoided.

opadron replied:
I promised myself I wasn't going to work on my day off, but I just wanted to say that, looking at your example, I was surprised to have found it quite reasonable.  I will be sure to apply this idea to the relevant parts of the code. Thanks!

waxlamp replied:
```javascript
dayOff.then(activity => {
  if (activity.work) {
    return Promise.reject(new Error('HEY! WATCH SOME NETFLIX OR SOMETHING!'));
  }
});
```

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+
+  const runPromise = (
+    D(actions.ensureScratchDirectory())
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+      outputs.output_path_1 = `${prefix}:${states[page.elements[0]].name}`;
+      outputs.output_path_2 = `${prefix}:${states[page.elements[1]].name}`;
+    })
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+    .then(() => (
+      analysisUtils.runTask(task, { inputs, outputs }, { title, maxPolls })
+    ))
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+      metaDataPromise,
+      runPromise
+    ])
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+
+    .then(([, meta, result]) => {
+      let { outputId1, outputId2 } = result;
+      return (
```
Remove parentheses

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+          loadModel(outputId1, ItemModel),
+          loadModel(outputId2, ItemModel)
+        ])
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+              )
+            ))
+          )
+
```
Remove blank line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+                meta[i],
+                ([k, v]) => new Promise((resolve, reject) => {
+                  item.addMetadata(
+                    k, v, () => resolve(),
```
Hard to read - put `k`, `v` on previous line

waxlamp commented on web_client/analysis-modules/feature-matching.jsx:
```diff
@@ -117,7 +117,115 @@ const computeMatch = (data, page) => {
+            ))
+          ))
+        ))
+
```
Remove blank line

waxlamp commented on server/task_specs/feature-match.py:
```diff
@@ -274,52 +274,52 @@ def main_metric(a, b):
+
+list1 = None
+with open(input_path_1, 'rU') as input1:
+    list1 = [
```
```python
extractor = ColumnExtractor if mode1 == 'c' else RowExtractor
with open(...) as input1:
    list1 = list(extractor(input1, dialect1))
```

opadron replied:
Good catch.

waxlamp replied:
Also, is the call to `list()` necessary here? Does using a class with a `next()` method always yield a generator expression?

waxlamp commented on server/ui_specs/feature-match.yml:
```diff
@@ -7,10 +7,14 @@ elements:
     name: Input File 1
     type: fileSelection
+    options:
+      onlyFileTypes: dataBlock
```
I think `allowFileTypes` would be a better name for this option.

waxlamp commented on server/ui_specs/iGPSe.yml:
```diff
@@ -16,7 +16,8 @@ elements:
     options:
-      onlyTypes: "^mrna$"
+      onlyFileTypes: dataBlock
+      onlyDataTypes: "^mrna$"
```
Same here - `allowDataTypes` instead of `onlyDataTypes`

opadron replied:
#54

opadron replied:
Let me elaborate: #54 has been created to track the issue outside of this PR.

waxlamp commented on web_client/utils/scatter.jsx:
```diff
@@ -0,0 +1,38 @@
+ * scatter elements in array among n consumers
+ */
+const scatter = (array, n, interleave=false) => {
+  let i;
```
Better to declare this right in the for loop:

```js
for(let i=0; i<n; i++)
```

waxlamp commented on web_client/utils/scatter.jsx:
```diff
@@ -0,0 +1,38 @@
+  let i;
+  let a, b=0;
+  let m = array.length;
+  let mModN = m%n;
```
The code will read easier if you don't store these in variables, and instead just compute them in place.

waxlamp replied:
If these stay, however, they should be declared `const`, not `let`.

waxlamp commented on web_client/utils/scatter.jsx:
```diff
@@ -0,0 +1,38 @@
+/**
+ * scatter elements in array among n consumers
+ */
+const scatter = (array, n, interleave=false) => {
```
May want to write some tests for this function.

waxlamp commented on web_client/components/containers/upload.jsx:
```diff
@@ -5,7 +5,7 @@ import Upload from '../body/upload';
 import { rest } from '../../globals';
 import objectReduce from '../../utils/object-reduce';
-import { Promise } from '../../utils/promise';
+import { Promise, SCHEDULES, map as promiseMap } from '../../utils/promise';
```
Avoid all-caps variable names.

waxlamp commented on web_client/components/containers/upload.jsx:
```diff
@@ -5,7 +5,7 @@ import Upload from '../body/upload';
 import { rest } from '../../globals';
 import objectReduce from '../../utils/object-reduce';
-import { Promise } from '../../utils/promise';
+import { Promise, SCHEDULES, map as promiseMap } from '../../utils/promise';
```
Why isn't it just called `promiseMap` to begin with?

Alternatively, you could put `map` on `Promise`'s prototype.

opadron replied:
The "map" name is the most frank and straightforward description for the function while still being unique within the namespace in the promise utility module.  This is the same rule-of-thumb I use when naming most other exported members.  It just so happens that there are plenty of other "map"s (including `Promise.map`, which I'd rather not monkey patch over).

I suspect that in most cases, client modules importing the symbol will choose to give it a more descriptive name within their namespace, but that isn't enough reason to preempt them in this regard.  Indeed, client modules could choose to import all the symbols like so:

    import * as promiseUtils

Clearly in this case, `map` is the ideal name, as it's resulting fully qualified name remains descriptive without being redundant.

This question could just as easily be one of why the symbols are imported the way they are.  My reasoning on this matter boils down to the fact that `Promise` is used everywhere, so I could either type `promiseUtils.Promise` ad nauseam, import the symbols in two steps (`import ... ; let Promise = ...;`), or actually take advantage of an ES6 language feature for its intended purpose.  I figured the latter was the best among admittedly sub-optimal options.

waxlamp replied:
I agree that monkeypatching `Promise.map` is probably a bad idea.

I've been bitten in other projects by bad name clashes that weren't so easily fixable with renaming imports as in ES6. That's the main reason for my caution here. But if we want to support this workflow by the client of renaming things from the most accurate description to something that doesn't clash, that's ok with me.

waxlamp commented on web_client/utils/promise.jsx:
```diff
@@ -1,4 +1,155 @@
+  STATIC: 'static'
+};
+
+export const map = (args) => {
```
In order to fit in with standard map-type functions, I feel like this function should be called as `map(array, mapper, opts)`, where opts would contain `factor` and `schedule`.

waxlamp replied:
Also, why not destructure `args` directly, including specifying the default values?

waxlamp commented on web_client/utils/promise.jsx:
```diff
@@ -1,4 +1,155 @@
+
+  callback(newInc);
+});
+
```
Should probably write some tests for these functions.

waxlamp commented on web_client/utils/load-model.jsx:
```diff
@@ -0,0 +1,16 @@
+      e.payload = payload;
+      reject(e);
+    });
+    req.then(() => { resolve(result); });
```
Remove curly brackets

opadron replied:
I don't actually want to return the value of `resolve(result)` (whatever it is).

waxlamp replied:
Ah, gotcha. In that case, I'd recommend breaking the lines so it doesn't get mistaken for a "short" arrow function:

```javascript
req.then(() => {
  resolve(result);
});
```

waxlamp reviewed (commented):
Mostly style comments, some substance suggestions. Take a look and let me know what you think.

ronichoudhury commented:
If you don't want to implement the style feedback here, that should be ok but I do nonetheless suggest we incorporate linting as a standard check or something for PRs.

Commit:
extend server-side preprocessing for task specs/factor common task code out to their own file

Commit:
fix javascript linting errors

Commit:
fix python linting errors/refactor common task spec code

Commit:
update travis (linting only, for now)

opadron commented on web_client/actions.jsx:
```diff
@@ -1,18 +1,20 @@
-import { isNil, isString, isUndefined } from 'lodash'; import URI from 'urijs';
```
FYI no idea how or when that line got that way!

Commit:
add most style changes suggested by @ronichoudhury

Commit:
change promiseUtils.map api to more closely resemble other mapping function signatures

Commit:
remind travis to run npm install

Commit:
add requirements.txt file

opadron commented:
@ronichoudhury PTAL.  I have made substantial improvements as a result of your feedback.  Travis testing is back, but currently limited to linting.  I've fixed all linting issues, except for one complexity issue raised for the `runTask` route, and I have added or addressed almost all of your comments.  Some of my responses prompt for a second look, so please take another pass and let me know if you have another round of changes to suggest/questions to ask.

opadron commented:
Also, with the latest commits, I'd like to propose that we try to leave styling suggestions to the linters, except for especially troubling cases.  While we should continue to work together on refining our style configuration, and expand on the situations that we'd like the linters to catch, I think we should try to get out of the business of bike shedding over parentheses and whitespace.  Again, for special cases, we can go back to handling them as-needed.

A lot of little nitpicks, like we had in this PR, definitely add up, and I want to make sure that next time, we're really getting something concrete out of that extra work.

Even then, this PR is ultimately better for it all, and I don't mean to diminish the value of @ronichoudhury's contributions, here.

ronichoudhury commented on server/task_specs/feature-apply.py:
```diff
@@ -0,0 +1,147 @@
+
```
Sorry, which docstring do you mean?

Commit:
break up short function into multiple lines
